File: Java/src/session1/td/NatParInt.java
package session1.td;

public class NatParInt implements Nat{
	//attributs
	public static FabriqueNat<Nat> FAB = new NatParInt(0);
	
	private int val;
	
	//constructor
	NatParInt(int val){
		if(val>=0) {
			this.val = val;
		}else {
			throw new java.lang.IllegalArgumentException();
		}
	}
	
	//methodes
	
	@Override
	public int val() {
		return this.val;
	}

	@Override
	public Nat creerNatAvecValeur(int x) {
		return new NatParInt(x);
	}

	@Override
	public boolean estNul() {
		return this.val()==0;
	}

	@Override
	public Nat predecesseur() {
		if(this.estNul()) {
			throw new java.lang.UnsupportedOperationException();
		}else {
			return new NatParInt(this.val()-1);
		}
		
	}
	
	@Override
	public String toString(){
		return String.valueOf(this.val());
	}
	@Override
	public boolean equals(Object o) {
		//System.out.println("o.val()"+ o.val());
		//System.out.println("this.val()"+this.val());
		return ((NatParInt)o).val()==this.val();
	}

	@Override
	public Nat creerZero() {
		return this.creerNatAvecValeur(0);
		//return new Zero();
	}

	@Override
	public int taille() {		
		return String.valueOf(this.val()).length();
	}

	@Override
	public int chiffre(int i) {
		String s = String.valueOf(this.val());
		return Integer.parseInt(String.valueOf(s.charAt(i)));
	}

	@Override
	public Nat somme(Nat n) {
		try {
			int result = this.val() + n.val();
			return this.creerNatAvecValeur(result);
		} catch (Exception e) {
			System.out.println(e.getMessage());
			return this.creerZero();
		}
		
	}

	@Override
	public Nat creerSuccesseur(Nat predecesseur) {
		return this.creerNatAvecValeur(predecesseur.val()+1);
	}

	@Override
	public Nat modulo(Nat x) {
		return this.creerNatAvecValeur(this.val() % x.val());
	}

	@Override
	public Nat div(Nat x) {
		return this.creerNatAvecValeur(this.val()/x.val());
	}

	@Override
	public Nat zero() {
		return this.creerZero();
	}

	@Override
	public Nat produit(Nat n) {
		return this.creerNatAvecValeur(this.val() * n.val());
	}

	@Override
	public Nat un() {
		return this.creerNatAvecValeur(1);
	}

	
}

File: Typescript/session1.td/NatParInt.ts
import { Nat } from "./Nat";
import { FabriqueNat } from "./FabriqueNat";

export class NatParInt implements Nat{
    public static FAB:FabriqueNat<Nat>=new NatParInt(0);

    constructor(private value:number){
        if(value>=0){
            this.value=value;
        } else {
            throw new Error("Le nombre n'est pas un entier naturel");
        }
    }

    getVal(): number {
        return this.value;
    }
    estNul(): boolean {
        if(this.val()==0){
            return true;
        } else if (this.val()>0){
            return false;
        } else {
            throw new Error("Le nombre n'est pas un entier naturel");
        }
    }    
    predecesseur(): Nat {
        if(this.estNul()){
            throw new Error("Le predecesseur n'est pas un entier naturel");
        } else {
            return new NatParInt(this.val()-1);
        }
    }
    chiffre(value: number): number {
        var s = String(this.val());
        return Number(String(s.charAt(value)));
    }
    taille(): number {
        return String(this.val()).length;
    }
    val(): number {
        return this.getVal();
    }
    creerNatAvecValeur(value: number): NatParInt {
        if (value<0){
            throw new Error("Le nombre n'est pas un entier naturel");
        } else {
            return new NatParInt(value);
        }
    }
    creerZero(): Nat {
        return new NatParInt(0);
    }
    creerSuccesseur(nat: Nat): Nat {
        return new NatParInt(nat.val()+1);
    }
    creerNatAvecRepresentation(chaine: string): Nat {
        if(Number(chaine)<0){
            throw new Error("Le nombre n'est pas un entier naturel");
        } else {
            return this.creerNatAvecValeur(Number(chaine));
        }
    }
    somme(x: Nat): NatParInt {
        return new NatParInt(this.val() + x.val());
    }
    zero(): Nat {
        return this.creerZero();
    }
    produit(x: Nat): Nat {
        return new NatParInt(this.val()*x.val());
    }
    un(): Nat {
        return this.creerNatAvecValeur(1);
    }
    modulo(x: Nat): Nat {
        if(x.val()!=0){
            return new NatParInt(this.val() % x.val())
        } else {
            throw new Error("Division par zéro");
        }
    }
    div(x: Nat): Nat {
        if(x.val()!=0){
            return new NatParInt(this.val() / x.val());
        } else {
            throw new Error("Divison par zéro");
        }
    }

    toString(): string {
        return String(this.val());
    }  

    equals(nat : Nat): boolean {
        if(nat.val() == this.val()){
            return true;
        } else {
            return false;
        }
    }



}